### PR TITLE
Users who vote >10x in 3min on the same user get a warning popup

### PR DESCRIPTION
**Pull request by jimrandomh**
Users who vote on content by the same user more than ten times in a 3-minute interval will get a popup that looks like this:

Screenshot 2023-01-17 at 16 44 22

This is rate-limited to trigger no more than once per hour. This will also add an automod notice to the moderator notes.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203769498070629) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/voteServer.ts [hunk 1]
```diff
@@ -234,39 +258,105 @@ export const performVoteServer = async ({ documentId, document, voteType, extend
...
+    type: RECEIVED_VOTING_PATTERN_WARNING,
+  }, {
+    sort: {createdAt: -1}
+  });
+  if (!mostRecentWarning) {
+    return false;
+  }
+  const warningAgeMS = new Date().getTime() - mostRecentWarning.createdAt.getTime()
+  const warningAgeMinutes = warningAgeMS / (1000*60);
+  return warningAgeMinutes < 60;
```
Comment (jpaddison3) [inline packages/lesswrong/server/voteServer.ts:279]: That's quite recent. ¯\\\_(ツ)_/¯ Sure.

File: packages/lesswrong/server/voteServer.ts [hunk 2]
```diff
@@ -234,39 +258,105 @@ export const performVoteServer = async ({ documentId, document, voteType, extend
...
+  perHour: number,
+  perUserPerDay: number
+}
+interface VotingRateLimit {
+  voteCount: number
+  periodInMinutes: number
+  types: "all"|"onlyStrong"|"onlyDown"
+  users: "allUsers"|"singleUser"
+  consequences: ("warningPopup"|"denyThisVote")[]
+  // TODO Consequences to add, not yet implemented: blockVotingFor24Hours, flagForModeration, revertRecentVotes
```
Comment (jpaddison3) [inline packages/lesswrong/server/voteServer.ts:293]: The first two do seem quite nice

File: packages/lesswrong/server/voteServer.ts [hunk 3]
```diff
@@ -234,39 +258,105 @@ export const performVoteServer = async ({ documentId, document, voteType, extend
...
+
+const getVotingRateLimits = (user: DbUser): VotingRateLimit[] => {
   if (user?.isAdmin) {
-    // Very lax rate limiting for admins
-    return {
-      perDay: 100000,
-      perHour: 50000,
-      perUserPerDay: 50000
-    }
+    return [];
```
Comment (jpaddison3) [inline packages/lesswrong/server/voteServer.ts:299]: Muahaha

File: packages/lesswrong/server/voteServer.ts [hunk 4]
```diff
@@ -234,39 +258,105 @@ export const performVoteServer = async ({ documentId, document, voteType, extend
...
+  } else {
+    return [
+      {
+        voteCount: 200,
+        periodInMinutes: 60,
```
Comment (jpaddison3) [inline packages/lesswrong/server/voteServer.ts:304]:
```suggestion
        periodInMinutes: 60 * 24,
```

File: packages/lesswrong/server/voteServer.ts [hunk 5]
```diff
@@ -234,39 +258,105 @@ export const performVoteServer = async ({ documentId, document, voteType, extend
...
+}
+
+interface VotingRateLimitSet {
+  perDay: number,
+  perHour: number,
+  perUserPerDay: number
+}
+interface VotingRateLimit {
+  voteCount: number
+  periodInMinutes: number
```
Comment (jpaddison3) [inline packages/lesswrong/server/voteServer.ts:289]:
```suggestion
  /** Must be ≤ than 24 hours */
  periodInMinutes: number
```

**Comment by jpaddison3** [comment]
1. I've added a relevant tiny type update, hope you don't mind.
2. This seems begging for some tests. Could I interest you in taking the test writing infrastructure for a spin?

Other than that I have one important suggest-edit comment, and one moderate one, and that's it.

**Review by jpaddison3 — APPROVED** [review]
LGTM, really nice work

**Comment by jpaddison3** [comment]
Now waiting on @darkruby501 to review wording for LW.

**Comment by darkruby501** [comment]
Wording LGTM

**Comment by jimrandomh** [comment]
I added a Cypress test `cypress/e2e/voting/vote_on_post.cy.js` which tests voting on a post, seeing its score update and seeing its author's karma update. It doesn't hit the rate limits (that sort of thing is a pain in the context of Cypress) but should be a decent step in that direction.

**Comment by jpaddison3** [comment]
@jimrandomh  I was thinking this would be a good db-integrated-but-not-browser-integrated[^1] test. But thanks a bunch for adding that test! It seems very useful to me!

[^1]: Integration test in our parlance. See `jest.integration.config.ts`. See also `packages/lesswrong/integrationTests/voting.tests.ts`.

**Pull request merged**